drivers: ethernet: stm32: remove phy specific code

maass-hamburg opened the pull request:
- enable mdio also on legacy devices
- remove phy specific code from stm32 ethernet driver
- setting the phy-handle of the ethernac mac is now mandatory
- phys, that don't use mdio or the standard registers can now be used.

maass-hamburg committed:
drivers: mdio: stm32: add mdio for legacy api

add mdio for legacy stm32 api

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
drivers: mdio: stm32: remove unused functions

The api takes care, if the mdio bus_enable and bus_disable
are not needed, so there is no need to have tis in the driver.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
drivers: ethernet: stm32: remove internal phy code

rewirte code and remove code that use internal phy functions.
A few Kconfig options got removed, that are now set by the
phy via the DT.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

marwaiehm-st commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -1290,36 +1121,98 @@ static int eth_init_api_v2(const struct device *dev)
-#if defined(CONFIG_ETH_STM32_AUTO_NEGOTIATION_ENABLE)
-	/* Auto Nego enabled */
-	get_auto_nego_speed_duplex(heth, &mac_config);
+	mac_config->DuplexMode =
```
The mac_config its not a pointer `mac_config.DuplexMode`

marwaiehm-st commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -1290,36 +1121,98 @@ static int eth_init_api_v2(const struct device *dev)
-	/* Auto Nego enabled */
-	get_auto_nego_speed_duplex(heth, &mac_config);
+	mac_config->DuplexMode =
+		PHY_LINK_IS_FULL_DUPLEX(state) ? ETH_FULLDUPLEX_MODE : ETH_MODE_HALFDUPLEX;
```
PHY_LINK_IS_FULL_DUPLEX(state->speed)

marwaiehm-st commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -1290,36 +1121,98 @@ static int eth_init_api_v2(const struct device *dev)
-									     : ETH_FULLDUPLEX_MODE;
-	mac_config.Speed = IS_ENABLED(CONFIG_ETH_STM32_SPEED_10M) ? ETH_SPEED_10M : ETH_SPEED_100M;
-#endif
+	mac_config->Speed = PHY_LINK_IS_SPEED_100M(state) ? ETH_SPEED_100M : ETH_SPEED_10M;
```
Ditto

marwaiehm-st commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -1290,36 +1122,98 @@ static int eth_init_api_v2(const struct device *dev)
-	/* Auto Nego enabled */
-	get_auto_nego_speed_duplex(heth, &mac_config);
+	mac_config.DuplexMode =
+		PHY_LINK_IS_FULL_DUPLEX(state->speed) ? ETH_FULLDUPLEX_MODE : ETH_MODE_HALFDUPLEX;
```
```suggestion
		PHY_LINK_IS_FULL_DUPLEX(state->speed) ? ETH_FULLDUPLEX_MODE : ETH_HALFDUPLEX_MODE;
```

mathieuchopstm commented on boards/st/stm32f746g_disco/stm32f746g_disco.dts:
```diff
@@ -177,17 +177,28 @@ zephyr_udc0: &usbotg_fs {
+	phy: ethernet-phy@0 {
+		compatible = "ethernet-phy";
+		reg = <0x00>;
+		status = "okay";
```
Nit: the default value for `status` is `"okay"` so this could be left out.

Ditto for all boards introducing the PHY as new node.

mathieuchopstm commented on drivers/ethernet/Kconfig.stm32_hal:
```diff
@@ -74,34 +74,6 @@ config ETH_STM32_HAL_USE_DTCM_FOR_DMA_BUFFER
-	help
-	  The phy address to use.
-
-config ETH_STM32_CARRIER_CHECK
-	bool "Check PHY carrier periodically"
-	default y
-	help
-	  Enables or disables checking of the PHY's carrier status.
-	  See also CONFIG_ETH_STM32_CARRIER_CHECK_RX_IDLE_TIMEOUT_MS
-
-config ETH_STM32_CARRIER_CHECK_RX_IDLE_TIMEOUT_MS
-	int "Carrier check timeout period (ms)"
-	depends on ETH_STM32_CARRIER_CHECK
-	default 500
-	range 100 30000
-	help
-	  Set the RX idle timeout period in milliseconds after which the
-	  PHY's carrier status is re-evaluated.
```
If I understand correctly, we no longer need this because the PHY driver should receive an IRQ when the PHY has something to say, and call back to us to signal carrier state?

maass-hamburg replied:
Yes, there is a callback from the phy, that is executed, when the link state changes(Mainly link up/down and speed). With that this Ethernet driver gets notified. This way we could also use phys, that are not connected via mdio.

mathieuchopstm commented on drivers/mdio/mdio_stm32_hal.c:
The Ethernet IP clock is not enabled by this driver? But what if it called before the STM32 ETH driver has started?

maass-hamburg replied:
At least for the STM32F7 I checked. the mdio controller uses the APB bus/HCLK as the clock and not the ethernet specific clocks

mathieuchopstm replied:
Indeed, only the bus clock is required to get the MDIO controller working... but that still means the bus clock needs to be enabled, right? :smile: 

c.f. https://github.com/mathieuchopstm/zephyr/blob/964421e574fdced52554a9016f0b95855e149525/dts/arm/st/f7/stm32f745.dtsi#L93-L100 and https://github.com/mathieuchopstm/zephyr/blob/49525fc20f2ae3e32f23aa5bd48cab744c157dde/drivers/mdio/mdio_ethbuiltin_stm32.c#L136-L143

This bug isn't easy to spot because most ST boards will work fine even if the PHY is not configured. If the clock is not enabled, the driver will appear to work (unfortunate consequence of "unclocked peripherals read as 0" combined with "the BUSY flag is active-high"), but in reality it won't do anything until the ETH driver starts and enables clocks.
There is one hint that something fishy is going on you should see on your side: if you use the generic `phy_mii` driver, it *should* log `PHY (0) ID 0` during boot, despite the PHY's ID definitely NOT being zero.

FTR: Instead of duplicating the `clocks` property, it could be placed on the parent `ethernet@XXX` node and consumed by both MAC and MDIO driver - I don't know which option is better though...

maass-hamburg replied:
you are right, I changed it

mathieuchopstm commented on dts/arm/st/f1/stm32f107.dtsi:
```diff
@@ -19,16 +19,25 @@
+				interrupts = <61 0>;
+				clock-names = "stmmaceth", "mac-clk-tx",
+					"mac-clk-rx";
+				clocks = <&rcc STM32_CLOCK(AHB1, 14U)>,
```
Nit, if you don't mind doing the changes :sweat_smile: : since we're touching the files, we might as well clean them up 
```suggestion
				clocks = <&rcc STM32_CLOCK(AHB1, 14)>,
```

Ditto for all similar lines.
(Rationale: https://github.com/zephyrproject-rtos/zephyr/pull/79683#pullrequestreview-2366609352)

marwaiehm-st commented on dts/arm/st/f1/stm32f107.dtsi:
```diff
@@ -19,16 +19,25 @@
+				interrupts = <61 0>;
+				clock-names = "stmmaceth", "mac-clk-tx",
+					"mac-clk-rx";
+				clocks = <&rcc STM32_CLOCK(AHB1, 14U)>,
+					<&rcc STM32_CLOCK(AHB1, 15U)>,
+					<&rcc STM32_CLOCK(AHB1, 16U)>;
+				status = "disabled";
```
```suggestion
				clocks = <&rcc STM32_CLOCK(AHB1, 14U)>,
					 <&rcc STM32_CLOCK(AHB1, 15U)>,
					 <&rcc STM32_CLOCK(AHB1, 16U)>;
				status = "disabled";
```

marwaiehm-st commented on dts/arm/st/h5/stm32h5.dtsi:
```diff
@@ -538,15 +538,17 @@
+				clock-names = "stmmaceth", "mac-clk-tx", "mac-clk-rx";
+				clocks = <&rcc STM32_CLOCK(AHB1, 19U)>,
+					<&rcc STM32_CLOCK(AHB1, 20U)>,
+					<&rcc STM32_CLOCK(AHB1, 21U)>;
```
Ditto

marwaiehm-st commented on dts/arm/st/h7/stm32h7.dtsi:
```diff
@@ -1025,15 +1025,17 @@
+				clock-names = "stmmaceth", "mac-clk-tx", "mac-clk-rx";
+				clocks = <&rcc STM32_CLOCK(AHB1, 15U)>,
+					<&rcc STM32_CLOCK(AHB1, 16U)>,
+					<&rcc STM32_CLOCK(AHB1, 17U)>;
```
Ditto

marwaiehm-st commented on dts/arm/st/n6/stm32n6.dtsi:
```diff
@@ -643,17 +643,19 @@
+					"mac-clk-rx";
+				clocks = <&rcc STM32_CLOCK(AHB5, 22)>,
+					<&rcc STM32_CLOCK(AHB5, 23)>,
+					<&rcc STM32_CLOCK(AHB5, 24)>;
```
Ditto

marwaiehm-st commented on dts/arm/st/f1/stm32f107.dtsi:
```diff
@@ -19,16 +19,25 @@
+			mac: ethernet {
+				compatible = "st,stm32-ethernet";
+				interrupts = <61 0>;
+				clock-names = "stmmaceth", "mac-clk-tx",
+					"mac-clk-rx";
```
```suggestion
				clock-names = "stmmaceth", "mac-clk-tx",
					      "mac-clk-rx";
```

marwaiehm-st commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -1288,38 +1117,104 @@ static int eth_init_api_v2(const struct device *dev)
 	}
+#else  /* CONFIG_ETH_STM32_HAL_API_V2 */
+	heth->Init.DuplexMode =
+		PHY_LINK_IS_FULL_DUPLEX(state->speed) ? ETH_MODE_FULLDUPLEX : ETH_HALFDUPLEX_MODE;
```
`ETH_MODE_HALFDUPLEX`  should be used on the API V1
```suggestion
		PHY_LINK_IS_FULL_DUPLEX(state->speed) ? ETH_MODE_FULLDUPLEX : `ETH_MODE_HALFDUPLEX`;
```

marwaiehm-st commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -1041,6 +968,17 @@ static int eth_initialize(const struct device *dev)
+		phy_get_link_state(eth_stm32_phy_dev, &state);
+
+		heth->Init.DuplexMode = PHY_LINK_IS_FULL_DUPLEX(state.speed) ? ETH_MODE_FULLDUPLEX
+									     : ETH_HALFDUPLEX_MODE;
```
ETH_HALFDUPLEX_MODE is for the API V2
```suggestion
									     : ETH_MODE_HALFDUPLEX;
```

mathieuchopstm reviewed (commented):
FWIW, I had something similar in stock (but not upstreamed); I have published the branch here if you want to take a look: https://github.com/mathieuchopstm/zephyr/tree/stm32_ethernet_rework

This branch is more conservative in terms of changes to the ETH driver, but uses some macrobatics in the MDIO driver to share the `ETH_HandleTypeDef` with its parent controller (saves RAM as the structure is quite large, but inhibits usage as a general-purpose MDIO; maybe an in-between solution could be found).

Another change I did was rename all "MDIO"-related stuff to clearly emphasize that this driver is for the MDIO controller built into the Ethernet IP (as opposed to the dedicated "general-purpose" MDIO controllers found on STM32H7) - this could be appreciated, but is mostly cosmetic, so not that important.

mathieuchopstm commented on drivers/mdio/mdio_stm32_hal.c:
```diff
@@ -96,6 +96,10 @@ static int mdio_stm32_init(const struct device *dev)
 		return ret;
 	}
 
+#ifdef CONFIG_ETH_STM32_HAL_API_V2
+	HAL_ETH_SetMDIOClockRange(&dev_data->heth);
+#endif
```
This was deleted from the Ethernet driver one commit ago. It should have been moved here directly (either in that commit, or the first one that touched this driver).

maass-hamburg replied:
the removal of that in the eth driver is unrelated, because it is unneeded there. `HAL_ETH_Init()` already contains this function.

mathieuchopstm replied:
Good catch; in this case, please add a comment in here explaining just that.

(This makes me a bit worried about F1/F2 series, since they appear to do the same thing, but don't provide a dedicated function... we might need to implement it ourselves for these series, but I guess we can leave that out for now)

maass-hamburg committed:
drivers: mdio: stm32: make driver more independent

make stm32 mdio driver more independent

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
drivers: ethernet: stm32: make mac a child like the mdio node

mac and mdio are now on the same level, this way
phy-handle can be used.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
boards: dts: stm32: add mdio and phy node

add mdio and phy node to every stm32board that
supports ethernet.
Also set the phy-handle for every ethernet mac.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
drivers: mdio: stm32: set mdio clock for v1 api

unfortunatly HAL_ETH_SetMDIOClockRange() isn't available
in the V1 API, so I had to copy the parts from the hal.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -54,42 +54,12 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
-static const struct device *eth_stm32_phy_dev = DEVICE_PHY_BY_NAME(0);
-
-#endif
+#define ETH_STM32_AUTO_NEGOTIATION_ENABLE                                                          \
+	UTIL_NOT(DT_NODE_HAS_PROP(DT_INST_PHANDLE(0, phy_handle), fixed_link))
```
this is only used by the legacy api. Also this way it can be evaluated during build.

clamattia replied:
I believe there are also makros to check whether any instace has prop `fixed_link`. That could be used to eliminate code form the build. I do not remember what is what called now.

mathieuchopstm replied:
Driver doesn't support multi-instance, so there's no issue with using a global for now IMO.

clamattia replied:
Ok 👍

mathieuchopstm commented on drivers/mdio/mdio_stm32_hal.c:
```diff
@@ -73,27 +85,62 @@ static int mdio_stm32_write(const struct device *dev, uint8_t prtad,
+#else /* CONFIG_ETH_STM32_HAL_API_V1 */
+	uint32_t tmpreg1 = 0U
+	/* Get the ETHERNET MACMIIAR value */
+	tmpreg1 = (heth->Instance)->MACMIIAR;
+	/* Clear CSR Clock Range CR[2:0] bits */
+	tmpreg1 &= ETH_MACMIIAR_CR_MASK;
+
+	/* Get hclk frequency value */
+	hclk = HAL_RCC_GetHCLKFreq();
+
+	/* Set CR bits depending on hclk value */
+#ifdef SOC_SERIES_STM32F1X
+	if ((hclk >= 20000000U) && (hclk < 35000000U)) {
+		/* CSR Clock Range between 20-35 MHz */
+		tmpreg1 |= (uint32_t)ETH_MACMIIAR_CR_DIV16;
+	} else if ((hclk >= 35000000U) && (hclk < 60000000U)) {
+		/* CSR Clock Range between 35-60 MHz */
+		tmpreg1 |= (uint32_t)ETH_MACMIIAR_CR_DIV26;
+	} else {
+		/* CSR Clock Range between 60-72 MHz */
+		tmpreg1 |= (uint32_t)ETH_MACMIIAR_CR_DIV42;
+	}
+#else  /* SOC_SERIES_STM32F2X */
+	if ((hclk >= 20000000U) && (hclk < 35000000U)) {
+		/* CSR Clock Range between 20-35 MHz */
+		tmpreg1 |= (uint32_t)ETH_MACMIIAR_CR_Div16;
+	} else if ((hclk >= 35000000U) && (hclk < 60000000U)) {
+		/* CSR Clock Range between 35-60 MHz */
+		tmpreg1 |= (uint32_t)ETH_MACMIIAR_CR_Div26;
+	} else if ((hclk >= 60000000U) && (hclk < 100000000U)) {
+		/* CSR Clock Range between 60-100 MHz */
+		tmpreg1 |= (uint32_t)ETH_MACMIIAR_CR_Div42;
+	} else {
+		/* CSR Clock Range between 100-120 MHz */
+		tmpreg1 |= (uint32_t)ETH_MACMIIAR_CR_Div62;
+	}
+#endif /* SOC_SERIES_STM32F2X */
+
+	/* Write to ETHERNET MAC MIIAR: Configure the ETHERNET CSR Clock Range */
+	(heth->Instance)->MACMIIAR = (uint32_t)tmpreg1;
```
Also please something to ensure we use HALv2 on non-F1/F2, e.g.:
```diff
-  #endif /* SOC_SERIES_STM32F2X */
+  #else /* !SOC_SERIES_STM32F1X && !SOC_SERIES_STM32F2X */
+  BUILD_ASSERT(0, "HALv2 required for series other than STM32F1 / STM32F2");
+  #endif
```
(maybe it can be done via Kconfig instead)

maass-hamburg replied:
It is already ensured in the ethernet Kconfig.stm32_hal:
```
choice ETH_STM32_HAL_API_VERSION
	prompt "STM32Cube HAL Ethernet version"

config ETH_STM32_HAL_API_V2
	bool "Use official STM32Cube HAL driver"
	depends on SOC_SERIES_STM32H7X || SOC_SERIES_STM32H5X || SOC_SERIES_STM32F4X || SOC_SERIES_STM32F7X || SOC_SERIES_STM32N6X
	select USE_STM32_HAL_ETH_EX if SOC_SERIES_STM32N6X
	help
	  Use the official STM32Cube HAL driver instead of the legacy one.

config ETH_STM32_HAL_API_V1
	bool "Use legacy STM32Cube HAL driver"
	depends on SOC_SERIES_STM32F1X || SOC_SERIES_STM32F2X
	help
	  Driver version based on legacy HAL version as the current official API version.
	  Available only for STM32F1 and STM32F2 SoC series.

endchoice
```

maass-hamburg committed:
drivers: ethernet: mdio: stm32: move stmmaceth clock to parent

move stmmaceth clock to parent, so it can also be
used by mdio and rename it to ``stm-eth``.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

mathieuchopstm commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -54,42 +54,12 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
-
-#endif
+#define ETH_STM32_AUTO_NEGOTIATION_ENABLE                                                          \
+	UTIL_NOT(DT_NODE_HAS_PROP(DT_INST_PHANDLE(0, phy_handle), fixed_link))
```
[Per `DT_NODE_HAS_PROP`'s documentation](https://github.com/zephyrproject-rtos/zephyr/blob/ca0f5f6924cdc7046ce8dbb6d32e39125b5ded7f/include/zephyr/devicetree.h#L3771-L3785):
```
To get a boolean property's truth value, use DT_PROP(node_id, prop) instead.
```
```suggestion
	UTIL_NOT(DT_PROP(DT_INST_PHANDLE(0, phy_handle), fixed_link))
```

maass-hamburg replied:
from ethernet-phy.yaml:
```
  fixed-link:
    type: string
    description: This link is fixed and does not require PHY configuration
    enum:
      - "10BASE-T Half-Duplex"
      - "10BASE-T Full-Duplex"
      - "100BASE-T Half-Duplex"
      - "100BASE-T Full-Duplex"
      - "1000BASE-T Half-Duplex"
      - "1000BASE-T Full-Duplex"
```
it is not a bool

mathieuchopstm replied:
Oh. Nevermind then 🙇

mathieuchopstm commented on dts/bindings/ethernet/st,stm32-ethernet-parent.yaml:
```diff
@@ -0,0 +1,16 @@
+# Copyright The Zephyr Project Contributors
+# SPDX-License-Identifier: Apache-2.0
+
+description: ST STM32 Ethernet parent node
+
+compatible: "st,stm32-ethernet-parent"
```
What do you think about renaming the existing `st,stm32<XXX>-ethernet` to `st,stm32<XXX>-ethernet-mac` then naming this binding/node:
```suggestion
description: STM32 Ethernet controller

compatible: "st,stm32-ethernet"
```

(`st,stm32-mdio` could also be renamed `st,stm32-ethernet-mdio` but that's cosmetic)

maass-hamburg replied:
I'm unsure, I'm more on the side of not renaming too much.

For mdio, I didn't find a separate mdio controller in the datasheets and hal, what I instead found was a MDIO slave interface, so it can act as a peripheral on the mdio bus. Currently this functionality is not supported by the zephyr mdio api. And for DT compatible, that should then reflect, that it is not a controller by a describing suffex.

mathieuchopstm replied:
You're correct again, I misunderstood RM0399 §57 as MDIO master... ~~in hindsight, everything makes much more sense when you consider the IP is a MDIO slave 😅~~

My main dislike point is the `ethernet-parent` name... Would renaming this binding to `st,stm32-ethernet-controller` - without changing the other bindings - be acceptable for you?
(We could add a note in the `st,stm32<XXX>-ethernet` bindings' description indicating they are for the MAC)

marwaiehm-st commented on drivers/mdio/mdio_stm32_hal.c:
```diff
@@ -73,27 +88,73 @@ static int mdio_stm32_write(const struct device *dev, uint8_t prtad,
+	/* The legacy V1 HAL API does not provide a way to set the MDC clock range
+	 * via a separated function call.
+	 */
+	uint32_t tmpreg1 = 0U
```
```suggestion
	uint32_t tmpreg1 = 0U;
```

marwaiehm-st commented on drivers/mdio/mdio_stm32_hal.c:
```diff
@@ -73,27 +88,73 @@ static int mdio_stm32_write(const struct device *dev, uint8_t prtad,
+	tmpreg1 &= ETH_MACMIIAR_CR_MASK;
+
+	/* Get hclk frequency value */
+	hclk = HAL_RCC_GetHCLKFreq();
```
The `hclk` undeclared

erwango commented on drivers/mdio/mdio_stm32_hal.c:
```diff
@@ -73,27 +88,74 @@ static int mdio_stm32_write(const struct device *dev, uint8_t prtad,
+	tmpreg1 &= ETH_MACMIIAR_CR_MASK;
+
+	/* Get hclk frequency value */
+	hclk = HAL_RCC_GetHCLKFreq();
```
Do you actually need hclk or bus freq that you can get with `clock_control_get_rate()` ?

maass-hamburg replied:
I copied this part straight from the hal, but that can be replaced with CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC

mathieuchopstm commented:
> Can someone verify, that it works on boards, that use the v1 eth hal api? I don't have a stm32f1* or stm32f2* board on hand.

Tested OK on Nucleo-F207ZG with `samples/net/dhcpv4_client`. I haven't found `stm3210c_eval` board here, and it seems to be the only STM32F1 board supported in Zephyr that has Ethernet, so F1 is untested for now...

Not 100% related but the call to `HAL_ETH_Init()` in `eth_initialize()` is blocking, delaying boot and console output by 5s when cable is unplugged (and still ~1.5s when plugged). I would have sworn I didn't encounter this behavior when I was testing my original patch, but I can't see any commit that changed this behavior, so I'm not really sure that my memory is correct... either way, this falls outside PR scope so ignore it.

mathieuchopstm commented on dts/arm/st/f1/stm32f107.dtsi:
```diff
@@ -19,16 +19,25 @@
-				 <&rcc STM32_CLOCK(AHB1, 16U)>;
-			status = "disabled";
+			clock-names = "stmmaceth";
+			clocks = <&rcc STM32_CLOCK(AHB1, 14)>;
+			mac: ethernet {
```
To improve readability, please add a newline between the properties and the child nodes:
```suggestion
			clocks = <&rcc STM32_CLOCK(AHB1, 14)>;

			mac: ethernet {
```
Ditto for all DTSI.

maass-hamburg replied:
done

maass-hamburg commented:
> > Can someone verify, that it works on boards, that use the v1 eth hal api? I don't have a stm32f1* or stm32f2* board on hand.
> 
> Tested OK on Nucleo-F207ZG with `samples/net/dhcpv4_client`. I haven't found `stm3210c_eval` board here, and it seems to be the only STM32F1 board supported in Zephyr that has Ethernet, so F1 is untested for now...
> 
> Not 100% related but the call to `HAL_ETH_Init()` in `eth_initialize()` is blocking, delaying boot and console output by 5s when cable is unplugged (and still ~1.5s when plugged). I would have sworn I didn't encounter this behavior when I was testing my original patch, but I can't see any commit that changed this behavior, so I'm not really sure that my memory is correct... either way, this falls outside PR scope so ignore it.

The thing is, that ``HAL_ETH_Init()`` in the v1 api also communicates with the phy via mdio. Best thing would be, that f1 and f2 get the new api in the hal, next best thing would be to copy the content of these functions to the zepyhr tree and remove the communication part with the phy.

mathieuchopstm commented on drivers/mdio/mdio_stm32_hal.c:
```diff
@@ -73,27 +88,76 @@ static int mdio_stm32_write(const struct device *dev, uint8_t prtad,
+	tmpreg1 = (heth->Instance)->MACMIIAR;
+	/* Clear CSR Clock Range CR[2:0] bits */
+	tmpreg1 &= ETH_MACMIIAR_CR_MASK;
+
+	/* Set CR bits depending on CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC value */
+#ifdef SOC_SERIES_STM32F1X
+	if ((CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC >= 20000000U) &&
+	    (CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC < 35000000U)) {
+		/* CSR Clock Range between 20-35 MHz */
+		tmpreg1 |= (uint32_t)ETH_MACMIIAR_CR_DIV16;
+	} else if ((CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC >= 35000000U) &&
+		   (CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC < 60000000U)) {
+		/* CSR Clock Range between 35-60 MHz */
+		tmpreg1 |= (uint32_t)ETH_MACMIIAR_CR_DIV26;
+	} else {
+		/* CSR Clock Range between 60-72 MHz */
+		tmpreg1 |= (uint32_t)ETH_MACMIIAR_CR_DIV42;
+	}
+#else  /* SOC_SERIES_STM32F2X */
+	if ((CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC >= 20000000U) &&
+	    (CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC < 35000000U)) {
+		/* CSR Clock Range between 20-35 MHz */
+		tmpreg1 |= (uint32_t)ETH_MACMIIAR_CR_Div16;
+	} else if ((CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC >= 35000000U) &&
+		   (CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC < 60000000U)) {
+		/* CSR Clock Range between 35-60 MHz */
+		tmpreg1 |= (uint32_t)ETH_MACMIIAR_CR_Div26;
+	} else if ((CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC >= 60000000U) &&
+		   (CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC < 100000000U)) {
+		/* CSR Clock Range between 60-100 MHz */
+		tmpreg1 |= (uint32_t)ETH_MACMIIAR_CR_Div42;
+	} else {
+		/* CSR Clock Range between 100-120 MHz */
+		tmpreg1 |= (uint32_t)ETH_MACMIIAR_CR_Div62;
+	}
+#endif /* SOC_SERIES_STM32F2X */
+
+	/* Write to ETHERNET MAC MIIAR: Configure the ETHERNET CSR Clock Range */
+	(heth->Instance)->MACMIIAR = (uint32_t)tmpreg1;
+#endif /* CONFIG_ETH_STM32_HAL_API_V1 */
```
Please split out to a dedicated function somewhere else in this file:
```suggestion
#ifdef CONFIG_ETH_STM32_HAL_API_V2
	HAL_ETH_SetMDIOClockRange(heth);
#else /* CONFIG_ETH_STM32_HAL_API_V1 */
	/* The legacy V1 HAL API does not provide a way to set the MDC clock range
	 * via a separated function call. Implement an equivalent function ourselves
	 * based on what the V1 HAL performs in HAL_ETH_Init().
	 */
        eth_set_mdio_clock_range_for_hal_v1(heth);
#endif /* CONFIG_ETH_STM32_HAL_API_V1 */
```

(function name is a suggestion, feel free to use something else if it feels better to you)

maass-hamburg replied:
sure

maass-hamburg replied:
done

maass-hamburg committed:
doc: update migration guide for stm32 ethernet driver

update migration guide for changes of the
stm32 ethernet driver.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

marwaiehm-st commented on drivers/mdio/mdio_stm32_hal.c:
```diff
@@ -102,22 +174,21 @@ static int mdio_stm32_init(const struct device *dev)
+	};                                                                                         \
+	static struct mdio_stm32_config mdio_stm32_config_##inst = {                               \
+		.pincfg = PINCTRL_DT_INST_DEV_CONFIG_GET(inst),                                    \
+		.pclken = {.bus = DT_CLOCKS_CELL_BY_NAME(DT_INST_PARENT(inst), stm_eth, bus),    \
+			   .enr = DT_CLOCKS_CELL_BY_NAME(DT_INST_PARENT(inst), stm_eth, bits)},  \
```
```suggestion
		.pclken = {.bus = DT_CLOCKS_CELL_BY_NAME(DT_INST_PARENT(inst), stm_eth, bus),    \
			   .enr = DT_CLOCKS_CELL_BY_NAME(DT_INST_PARENT(inst), stm_eth, bits)},  \
```
```suggestion
		.pclken = {.bus = DT_CLOCKS_CELL_BY_NAME(DT_INST_PARENT(inst), stm_eth, bus),      \
			   .enr = DT_CLOCKS_CELL_BY_NAME(DT_INST_PARENT(inst), stm_eth, bits)},    \
```

maass-hamburg commented:
> #85679 seems to be merged soon, have to apply changes to their new boards/st/nucleo_f439zi/nucleo_f439zi.dts too, when it's merged

Done

maass-hamburg commented:
works on ``stm32h573i_dk``

maass-hamburg commented:
During start the  error ``HAL_ETH_Stop{_IT} failed`` from line 1194 will be fired, when the phy is not up during phy callback set, because ``HAL_ETH_Stop_IT`` will be called before ``HAL_ETH_Start_IT`` (we can't stop, if it's already stopped); should I change it from a ``LOG_ERR`` to a ``LOG_DBG``, so users are not confused? Also that's the only case where the HAL will return a error on that function.

marwaiehm-st commented:
> During start the error `HAL_ETH_Stop{_IT} failed` from line 1194 will be fired, when the phy is not up during phy callback set, because `HAL_ETH_Stop_IT` will be called before `HAL_ETH_Start_IT` (we can't stop, if it's already stopped); should I change it from a `LOG_ERR` to a `LOG_DBG`, so users are not confused? Also that's the only case where the HAL will return a error on that function.


So ,The error message appears only on the board API v2. I guess It would be better to check the ETH state with HAL_ETH_GetState before stopping

```c
#if defined(CONFIG_ETH_STM32_HAL_API_V2)
hal_ret = HAL_ETH_GetState(heth);
if (hal_ret != HAL_ETH_STATE_STARTED) {
		LOG_DBG("ETH HAL driver is already stopped");
} else {
                // stopping
}
``` 
what do you think ?

mathieuchopstm commented:
> During start the error `HAL_ETH_Stop{_IT} failed` from line 1194 will be fired, when the phy is not up during phy callback set, because `HAL_ETH_Stop_IT` will be called before `HAL_ETH_Start_IT` (we can't stop, if it's already stopped); should I change it from a `LOG_ERR` to a `LOG_DBG`, so users are not confused? Also that's the only case where the HAL will return a error on that function.

I see that we call `net_eth_carrier_{on/off}` from the `phy_link_state_changed` callback, and these functions seem to keep track in the device which state we're in:
https://github.com/zephyrproject-rtos/zephyr/blob/fa45bebccfb1fa4ce25627518b0000cd80057555/subsys/net/l2/ethernet/ethernet.c#L920-L927

I would have said to use this to know whether we should stop the device, but it seems these functions may be no-op under certains circumstances.
https://github.com/zephyrproject-rtos/zephyr/blob/fa45bebccfb1fa4ce25627518b0000cd80057555/drivers/ethernet/eth_raw.c#L22-L25

As I assume the bookkeeping is not performed under these circumstances, and since the legacy HAL doesn't care about device state, I agree with @marwaiehm-st 's proposal - always call `HAL_ETH_Stop` on legacy, but check that HAL is started on API_V2 - if yes, perform `HAL_ETH_StopIT`; if no, `LOG_DBG` and do nothing.

mathieuchopstm commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -1288,38 +1116,105 @@ static int eth_init_api_v2(const struct device *dev)
+	hal_ret = HAL_ETH_Stop(heth);
+#endif
+	if (hal_ret != HAL_OK) {
+		/* Stopping the ETH HAL only fails, when it's already stopped. */
+		LOG_DBG("HAL_ETH_Stop{_IT} failed, already not running");
```
Let's avoid saying the function "failed" - rather, it hasn't done anything because it wasn't necessary.
```suggestion
		/* HAL_ETH_Stop{_IT} returns HAL_ERROR only if ETH is already stopped */
		LOG_DBG("HAL_ETH_Stop{_IT} returned error (Ethernet is already stopped)");
```

maass-hamburg replied:
applied

mathieuchopstm reviewed (dismissed):
LGTM in current state. Very minor nits in the log messages:

maass-hamburg commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -1038,9 +964,20 @@ static int eth_initialize(const struct device *dev)
+
 	hal_ret = HAL_ETH_Init(heth);
 	if (hal_ret == HAL_TIMEOUT) {
 		/* HAL Init time out. This could be linked to */
```
applied

maass-hamburg commented:
applied suggested changes by @mathieuchopstm
